### DISCUSSION TITLE
Last Line of a HTTP multipart formdata request with CRLF or not?

### DISCUSSION THREAD

**Discussion by mccare** [nodejs/undici#2186] [Q&A] [answered]
Hi,

I'm using undici to POST a multipart formdata to a webserver with https://github.com/yhirose/cpp-httplib as server library. The library expects every line to end in a CRLF. Undici does not send the last line with a CRLF (see [here](https://github.com/nodejs/undici/blob/e22448d8de7e8d051e9314d59774821cffd0f907/lib/fetch/body.js#L150)). So my question would be what is the correct behaviour? Add the CRLF to the last line of a http request or relax the checking in cpp-httplib to not require a CRLF on the last line?

Btw, curl does add a CRLF after the last line (after the boundary--). Maybe some expert could give me a hint before just "patching something"...

Thanks

Christian

**Comment by metcoder95** [comment, marked as answer]
Hi!

According to spec, `CRLF` seems to not be mandatory unless the body is of `form-data` content type, quoting:

>The message body is itself a protocol element and MUST therefore use only CRLF to represent line breaks between body-parts [RFC-2616](https://www.ietf.org/rfc/rfc2616.txt)

Although it just specifies it between body parts, not mentioning the last part of the body.
Just made quick look, haven't looked deeper into it yet

**Comment by mccare** [comment]
Thanks, did some digging and https://datatracker.ietf.org/doc/html/rfc2046#section-5.1 states the BNF form of a multipart body is:

```
multipart-body := [preamble CRLF]
                       dash-boundary transport-padding CRLF
                       body-part *encapsulation
                       close-delimiter transport-padding
                       [CRLF epilogue]
delimiter := CRLF dash-boundary
close-delimiter := delimiter "--"
```

So a CRLF is not required after the close-delimiter.
